frontend/src/features/recaptcha/useRecaptcha.tsx

```diff
@@ -0,0 +1,180 @@
+  theme?: 'light' | 'dark'
+  size?: 'compact' | 'normal' | 'invisible'
+  badge?: 'bottomright' | 'bottomleft' | 'inline'
+  tabindex?: number
+  hl?: string
+  isolated?: boolean
+}
+
+interface UseRecaptchaProps extends RecaptchaBaseConfig {
+  /** id of container to load reCaptcha in. If not provided, one will be generated */
+  id?: string
+  useRecaptchaNet?: boolean
+  useEnterprise?: boolean
+  onChange?: (response: string | null) => void
+  onExpiry?: () => void
+  onError?: () => void
+}
+
+type RecaptchaConfig = RecaptchaBaseConfig & {
+  callback?: (response: string) => void
+  'expired-callback'?: () => void
+  'error-callback'?: () => void
+}
+
+type GrecaptchaCallbacks = {
+  ready: (callback: () => void) => void
+  render: (container?: string, config?: RecaptchaConfig) => number
+  reset: (id?: number) => void
+  execute: (id?: number) => void
+  getResponse: (id?: number) => string
+}
+
+export type Grecaptcha = GrecaptchaCallbacks & {
+  enterprise: GrecaptchaCallbacks
+}
+
+declare global {
+  interface Window {
+    grecaptcha?: Grecaptcha
+  }
+}
+
+const getRecaptchaUrl = ({
+  useEnterprise,
+  useRecaptchaNet,
+}: Pick<UseRecaptchaProps, 'useEnterprise' | 'useRecaptchaNet'>) => {
+  const hostname = useRecaptchaNet ? 'recaptcha.net' : 'www.google.com'
+  return `https://${hostname}/recaptcha/${
+    useEnterprise ? 'enterprise' : 'api'
+  }.js?render=explicit`
+}
+
+type GreptchaExecutionCallback = {
+  resolve?: (response: string | null) => void
+  reject?: (reason?: unknown) => void
+}
+
+const DEFAULT_CONTAINER_ID = 'recaptcha-container'
+
+export const useRecaptcha = ({
+  id: containerId = DEFAULT_CONTAINER_ID,
+  sitekey,
+  theme,
+  onChange,
+  onExpiry,
+  onError,
+  useEnterprise = true,
+  badge = 'inline',
+  size = 'invisible',
+  useRecaptchaNet,
+}: UseRecaptchaProps) => {
+  useScript(getRecaptchaUrl({ useEnterprise, useRecaptchaNet }))
+
+  const grecaptcha: GrecaptchaCallbacks | undefined = useEnterprise
+    ? window.grecaptcha?.enterprise
+    : window.grecaptcha
+
+  const [hasLoaded, setHasLoaded] = useState(grecaptcha?.render ? true : false)
+  const [widgetId, setWidgetId] = useState<number>()
+
+  const executionPromise = useRef<GreptchaExecutionCallback>({})
+
+  useInterval(
+    () => {
+      if (grecaptcha?.render) {
+        setHasLoaded(true)
+      }
+    },
+    // Do not even start if sitekey is not provided
+    !sitekey || hasLoaded ? null : 500,
+  )
+
+  const handleChange = useCallback(
+    (response: string | null) => {
+      onChange?.(response)
+      if (executionPromise.current.resolve) {
+        executionPromise.current.resolve(response)
+        executionPromise.current = {}
+      }
+    },
+    [onChange],
+  )
+
+  const handleError = useCallback(() => {
+    onError?.()
+    if (executionPromise.current?.reject) {
+      executionPromise.current.reject()
+      executionPromise.current = {}
+    }
+  }, [onError])
+
+  const handleExpiry = useCallback(() => {
```

Comment:
as discussed, we can remove the `onXx` handlers as callers can just use the async version

Reply:
d8fd98d